src/charm.py

```diff
@@ -0,0 +1,169 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical
+# See LICENSE file for licensing details.
+#
+# Learn more at: https://juju.is/docs/sdk
+
+import logging
+import subprocess
+from pathlib import Path
+from typing import Optional
+
+import ops
+from charms.operator_libs_linux.v2.snap import SnapCache, SnapState
+from ops import (
+    ActionEvent,
+    ActiveStatus,
+    BlockedStatus,
+    ErrorStatus,
+    EventBase,
+    MaintenanceStatus,
+    WaitingStatus,
+)
+from ops.interface_kube_control import KubeControlRequirer
+from ops.interface_tls_certificates import CertificatesRequires
+
+logger = logging.getLogger(__name__)
+
+VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]
+
+
+def determine_arch() -> str:
+    """Dpkg wrapper to surface the architecture we are tied to."""
+    cmd = ["dpkg", "--print-architecture"]
+    output = subprocess.check_output(cmd).decode("utf-8")
+
+    return output.rstrip()
+
+
+class KubernetesE2ECharm(ops.CharmBase):
+    """Charm the service."""
+
+    CA_CERT_PATH = Path("/srv/kubernetes/ca.crt")
+
+    def __init__(self, *args) -> None:
+        super().__init__(*args)
+        self.kube_control = KubeControlRequirer(self)
+        self.certificates = CertificatesRequires(self)
+
+        self.CA_CERT_PATH.parent.mkdir(exist_ok=True)
+
+        self.snap_cache = SnapCache()
+
+        self.framework.observe(self.on.kube_control_relation_created, self._kube_control)
+        self.framework.observe(self.on.kube_control_relation_joined, self._kube_control)
+        self.framework.observe(self.on.kube_control_relation_changed, self._check_config)
+        self.framework.observe(self.on.kube_control_relation_broken, self._check_config)
```

Comment:
I think this is the preferred way now?
```suggestion
        kubecontrol = self.on["kube-control"]
        self.framework.observe(kubecontrol.relation_created, self._kube_control)
        self.framework.observe(kubecontrol.relation_joined, self._kube_control)
        self.framework.observe(kubecontrol.relation_changed, self._check_config)
        self.framework.observe(kubecontrol.relation_broken, self._check_config)
```